// devextreme/esm/ui/form/components/button_item.js
/**
 * DevExtreme (esm/ui/form/components/button_item.js)
 * Version: 21.1.6
 * Build date: Mon Sep 27 2021
 *
 * Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import $ from "../../../core/renderer";
import {
    isDefined
} from "../../../core/utils/type";
import {
    extend
} from "../../../core/utils/extend";
var FIELD_BUTTON_ITEM_CLASS = "dx-field-button-item";
export function renderButtonItemTo(_ref) {
    var {
        item: item,
        $container: $container,
        validationGroup: validationGroup,
        createComponentCallback: createComponentCallback,
        cssItemClass: cssItemClass
    } = _ref;
    $container.addClass(FIELD_BUTTON_ITEM_CLASS).css("textAlign", convertAlignmentToTextAlign(item.horizontalAlignment)).addClass(cssItemClass);
    $container.parent().css("justifyContent", convertAlignmentToJustifyContent(item.verticalAlignment));
    var $button = $("<div>");
    $container.append($button);
    return createComponentCallback($button, "dxButton", extend({
        validationGroup: validationGroup
    }, item.buttonOptions))
}

function convertAlignmentToTextAlign(horizontalAlignment) {
    return isDefined(horizontalAlignment) ? horizontalAlignment : "right"
}

function convertAlignmentToJustifyContent(verticalAlignment) {
    switch (verticalAlignment) {
        case "center":
            return "center";
        case "bottom":
            return "flex-end";
        default:
            return "flex-start"
    }
}
